elyra/cli/pipeline_app.py

```diff
@@ -311,5 +310,69 @@ def run(pipeline_path):
     print_banner("Elyra Pipeline Local Run Complete")
 
 
+@click.command()
+@click.option('--json',
+              'json_option',
+              is_flag=True,
+              required=False,
+              help='Returns pipeline file in a machine-readable JSON format')
+@click.argument('pipeline_path')
+def describe(json_option, pipeline_path):
+    """
+    Display pipeline description in a human readable format
+    """
+
+    _validate_pipeline_file_extension(pipeline_path)
+
+    pipeline_definition = _preprocess_pipeline(pipeline_path)
+
+    indent_length = 4
+    for current_pipeline in pipeline_definition["pipelines"]:
+        pipeline_keys = ["name", "description", "type", "version", "nodes", "dependencies"]
+
+        iter_keys = {"dependencies"}
+
+        describe_dict = OrderedDict()
+
+        properties = current_pipeline["app_data"]["properties"]
```

Comment:
This code fails for empty pipelines (pipelines without a node):

```
$ elyra-pipeline describe empty.pipeline 
Traceback (most recent call last):
  File "/opt/anaconda3/envs/cli-describe/bin/elyra-pipeline", line 8, in <module>
    sys.exit(pipeline())
  File "/opt/anaconda3/envs/cli-describe/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/opt/anaconda3/envs/cli-describe/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/opt/anaconda3/envs/cli-describe/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/anaconda3/envs/cli-describe/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/anaconda3/envs/cli-describe/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/opt/anaconda3/envs/cli-describe/lib/python3.7/site-packages/elyra/cli/pipeline_app.py", line 337, in describe
    properties = current_pipeline["app_data"]["properties"]
KeyError: 'properties'

```

Reply:
I'm currently testing empty pipelines using this pipeline with no nodes that recently got added to the `test_pipeline_app.py` file:
```
{
  "doc_type": "pipeline",
  "version": "3.0",
  "id": "0",
  "primary_pipeline": "1",
  "pipelines": [
    {
      "id": "1",
      "nodes": [],
      "app_data": {
        "runtime":"",
        "version": 4,
        "properties": {
          "name": "generic",
          "runtime": "Generic"
        }
      }, 
      "schemas": []
    }
  ]
}
```
This passes, so I'm not sure what an empty pipeline file actually looks like. If you could show me what the `empty.pipeline` you're testing with looks like, I shouldn't have a problem testing and resolving this on my own.

As a side note, if the CLI commands are expected to deal with a pipeline that's "more empty" than what the tests contain, the tests may be insufficient. All the current test pipelines in the `test_pipeline_app.py` do what I expect.

Reply:
The best way to produce an empty pipeline is to create a `.pipeline` file using the Visual Pipeline Editor in JupyterLab. Fwiw I used the Airflow pipeline editor. If you save that pipeline without adding any nodes you should be able to reproduce the scenario that wasn't working for me. (The issue is recreatable using any empty pipeline of any type generic/kfp/airflow.)

> As a side note, if the CLI commands are expected to deal with a pipeline that's "more empty" than what the tests contain, the tests may be insufficient.

In general the expectation would be that the code gracefully handles common [problem] scenarios:
 - input is not a file
 - input file is not valid JSON
 - input file does not contain the expected property names that the code relies on to produce the output

Reply:
> * input file does not contain the expected property names that the code relies on to produce the output

That explains a lot. Thanks for the help. I just pushed out some changes that make it impossible to run into key errors as long as `_preprocess_pipeline(pipeline_path)` always returns a `dict`.

I'm still going to make that empty file and test for its formatting, but it shouldn't cause an error anymore.

Reply:
Please note that I've opened #2066 to remove this "empty pipeline ambiguity" we're seeing.  This particular test artifact should be updated (not necessary for this PR - although that might prevent/eliminate possible "defensive" coding techniques) to reflect an "explicitly-saved" pipeline.